### PR TITLE
Fix #2032: avoid propagating mutated SROA fields

### LINKED ISSUE #2032: Follow up on #1981
Currently aggressive TDCU only works correctly, if no local variable is assigned in the aggregate as well as the body of the ILFunction. See `Test9` of the SROA test suite.

There are two options we already have discussed:
`original` and `copy` are `ILVariable`s
1) ensure that `original.LoadCount == 1 && original.IsSingeDefinition && stloc copy(ldloc original)  dominates all other uses of copy`
2) or neither original nor copy is mutated (after initialization), so all uses see the same reference: `original.IsSingleDefinition && copy.IsSingleDefinition && copy.StoreInstructions.Single().MatchLdLoc(original)`

Find a way to make sure either condition holds before and after TDCU or find another applicable condition.

### PR DESCRIPTION
**Pull request by siegfriedpammer**
## Summary
- Keep aggressive SROA from propagating a display-class field to its source variable when the field is assigned again after initialization.
- Re-enable the existing AggressiveScalarReplacementOfAggregates Test9 regression case.

Fixes #2032

## Testing
- OPENSSL_ENABLE_SHA1_SIGNATURES=1 dotnet test --project ICSharpCode.Decompiler.Tests/ICSharpCode.Decompiler.Tests.csproj --filter 'FullyQualifiedName~UglyTestRunner.AggressiveScalarReplacementOfAggregates' --report-trx

### CODE REVIEW

File: ICSharpCode.Decompiler/IL/Transforms/TransformDisplayClassUsage.cs [hunk 1]
```diff
@@ -227,6 +227,19 @@ bool ValidateUse(DisplayClass container, ILInstruction use)
...
+						{
+							// The field is mutated after initialization (stored again, or its address
+							// escapes). Propagation is only sound if the store can be redirected to
+							// the propagated variable: possible for parameters (their remaining uses
+							// are restricted in ResolveVariableToPropagate), but not for 'this' and
+							// not for a variable that is itself a scalar-replaced display class.
+							var propagatedVariable = variable.GetOrDeclare();
+							if (propagatedVariable.IsThis() || displayClasses.ContainsKey(propagatedVariable))
+							{
+								variable.Propagate(null);
```
Comment (christophwille) [inline ICSharpCode.Decompiler/IL/Transforms/TransformDisplayClassUsage.cs:240]: Cancelling propagation *during* the validation loop extends an existing order sensitivity: consider `dc2.f = dc1;` (dc1 itself a display class, so the field is propagatable) followed by a later mutation of `dc2.f`. If dc1 is validated before dc2, its store-into-field use is accepted via the `StObj` case on the strength of `declaredVar.CanPropagate == true`; dc2's validation then cancels that propagation here, leaving dc1 scalar-replaced while the surviving `dc2.f = dc1` store still references it. The pre-existing post-pass `Propagate(null)` had the same theoretical hazard, but this guard triggers in more situations. Aggressive-SROA-only and needs contrived input, so a follow-up (e.g. re-validating display classes referenced by a cancelled field's initializer, or iterating validation to a fixpoint) seems fine - just flagging it as a conscious decision.

File: ICSharpCode.Decompiler/IL/Transforms/TransformDisplayClassUsage.cs [hunk 2]
```diff
@@ -245,6 +258,23 @@ bool ValidateUse(DisplayClass container, ILInstruction use)
 			}
 		}
 
+		/// <summary>
+		/// True when the given field access is a plain read or one of the recorded
+		/// initializer stores; false for any other store or for an escaping address.
+		/// </summary>
+		static bool IsReadOnlyOrInitializerUse(LdFlda ldflda, VariableToDeclare variable)
```
Comment (christophwille) [inline ICSharpCode.Decompiler/IL/Transforms/TransformDisplayClassUsage.cs]: Nit: `IsReadOnlyOrInitializerUse` could be misread as relating to C# `readonly`. Something like `IsPlainReadOrInitializerStore` would be self-describing. Purely cosmetic, feel free to ignore.

File: ICSharpCode.Decompiler.Tests/TestCases/Ugly/AggressiveScalarReplacementOfAggregates.cs [hunk 3]
```diff
@@ -180,15 +181,36 @@ public void Test8b(int i)
...
+		public void Test10()
+		{
+			DisplayClass displayClass = new DisplayClass {
+				thisField = this,
+				field1 = 1,
+				field2 = "Hello World!"
+			};
+			Interlocked.Exchange(ref displayClass.thisField, new Program());
+			Console.WriteLine("{0} {1}", this, displayClass.thisField);
+		}
```
Comment (christophwille) [inline ICSharpCode.Decompiler.Tests/TestCases/Ugly/AggressiveScalarReplacementOfAggregates.cs:204]: Suggestion: add a fixture for the one propagate-despite-mutation path that still survives - a field initialized from a *non-this parameter* (single load) and then reassigned. That is now the only case where mutation and propagation coexist, and the soundness argument in the new `ValidateUse` comment rests on the `LoadCount == 1` restriction in `ResolveVariableToPropagate`, so it would be good to pin it down. Optionally also a variant where the mutation happens inside a nested lambda (should already be caught since `LoadInstructions` spans nested functions, but it is the classic display-class shape and cheap to lock in).

**Review by christophwille — COMMENTED** [review]
Reviewed both commits end to end, including a local run of the test suite.

**Summary**

1. The `ValidateUse` guard cancels propagation when a propagatable field is mutated after initialization (second store, or address escape like `Interlocked.Exchange`), but only when the propagation source is `this` or another display class. Non-this parameters keep propagating, which is sound because `ResolveVariableToPropagate` only allows them with `LoadCount == 1`, so the mutation store can be redirected to the parameter.
2. The `VisitStLoc` initializer-inlining change correctly skips initializer stores of still-propagated fields (previously they would re-emit as a self-assignment; for `this` that is invalid C#), and the `insertionIndex` bookkeeping keeps inserted stores contiguous. `context.EndStep(null)` is reachable when all stores are skipped, but `EndStep` is null-safe, so that is fine.

Details I checked: `IsReadOnlyOrInitializerUse` classifies all use shapes conservatively (nested `LdFlda`, `ldflda` as store value, call arguments all count as mutation); `Initializers` is non-empty whenever `CanPropagate` is true (both set together in `AddVariable`); `HandleInitBlock` breaks on duplicate fields so a second store in the init region is treated as mutation; the guard composes symmetrically with the existing post-pass in `AnalyzeFunction` that restricts surviving propagation to parameters and display classes; mutations inside nested lambdas are caught because `LoadInstructions` spans nested functions.

**Verification**

Ran `UglyTestRunner.AggressiveScalarReplacementOfAggregates` on this branch locally: all 4 `UseRoslyn4_14_0` configs pass. The 4 `UseRoslynLatest` configs failed in my environment, but an unrelated control test (`NoNewOfT`) fails identically (`csc` cannot load `System.Runtime.CompilerServices.Unsafe 6.0.3.0`), so that is a local toolset issue with the Roslyn 5.9.0 transport-feed compiler, not this PR.

**Verdict**

Looks good to me. The fix is conservative in the right places and well-commented. Inline comments cover one theoretical validation-order edge case and two optional test suggestions; none are blockers.

*This review was AI-assisted (Claude Code, claude-fable-5), driven and checked by a human.*

File: ICSharpCode.Decompiler/IL/Transforms/TransformDisplayClassUsage.cs (hunk 1 above)
Comment (siegfriedpammer) [reply ICSharpCode.Decompiler/IL/Transforms/TransformDisplayClassUsage.cs:240]: Tried hard to reproduce this and could not; sharing the evidence so the flag has a disposition.

1. **Targeted construction.** Built the exact scenario the comment describes (dc1 referenced from a recorded initializer store of dc2, later mutation of the field, second store to defeat the stloc rescue path). Tracing shows why it stays safe: the *container* is consistently validated first, so the cancellation lands before the referenced class is validated - it is then simply rejected and stays materialized, output correct.
2. **In-transform oracle over a corpus.** Instrumented the guard to report the exact hazardous sequence (cancelling propagation to a display class that already passed validation) and decompiled System.Private.CoreLib 8.0.29, System.Linq, System.Collections.Immutable, plus the repro assemblies from #3714 and #3202 (the latter obfuscated), under both SROA settings: zero firings across ~250k methods.
3. **Mechanism-guided inversion.** The protective order comes from initializer folding: a display class whose initializer folds into an object-initializer block is detected via a late-appended InitializerTarget temp, so the referenced class is enumerated after its container. A deliberately inverted construction (referenced class unfolded/detected early, container folded/detected late) still did not fire the oracle.

Not proven unreachable - the enumeration order is emergent, not enforced - so agreed on treating a fixpoint/re-validation as the follow-up shape if a trigger ever surfaces. No defensive code added now, deliberately: no failing fixture, nothing to exercise it.

_(Reply written by an AI agent on the author's behalf.)_

File: ICSharpCode.Decompiler/IL/Transforms/TransformDisplayClassUsage.cs (hunk 2 above)
Comment (siegfriedpammer) [reply ICSharpCode.Decompiler/IL/Transforms/TransformDisplayClassUsage.cs]: Renamed to `IsPlainReadOrInitializerStore` in 5a1115019.

_(Reply written by an AI agent on the author's behalf.)_

File: ICSharpCode.Decompiler.Tests/TestCases/Ugly/AggressiveScalarReplacementOfAggregates.cs (hunk 3 above)
Comment (siegfriedpammer) [reply ICSharpCode.Decompiler.Tests/TestCases/Ugly/AggressiveScalarReplacementOfAggregates.cs:204]: Added in 5a1115019: Test12 pins the surviving parameter path (output shows the store redirected to the single-load parameter), Test13 records the lambda variant - with a comment noting it does not exercise the guard, since capturing the display class keeps it materialized and propagation never arises there. Test13 routes the delegate through a helper because the Ugly runner defines no OPT symbol, so the Debug/Release lambda-inlining split cannot be handled with #if in these fixtures.

_(Reply written by an AI agent on the author's behalf.)_
